port/stack_trace.cc

```diff
@@ -51,28 +55,40 @@ namespace port {
 
 namespace {
 
-#if defined(OS_LINUX) || defined(OS_FREEBSD) || defined(OS_GNU_KFREEBSD)
+#if defined(OS_LINUX) || defined(OS_FREEBSD) || defined(OS_OPENBSD) || defined(OS_GNU_KFREEBSD)
 const char* GetExecutableName() {
   static char name[1024];
 
-#if !defined(OS_FREEBSD)
-  char link[1024];
-  snprintf(link, sizeof(link), "/proc/%d/exe", getpid());
-  auto read = readlink(link, name, sizeof(name) - 1);
-  if (-1 == read) {
+#if defined(OS_FREEBSD)
+  int mib[4] = {CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, -1};
+  size_t namesz = sizeof(name);
+
+  auto ret = sysctl(mib, 4, name, &namesz, nullptr, 0);
+  if (-1 == ret) {
     return nullptr;
   } else {
-    name[read] = 0;
     return name;
   }
-#else
-  int mib[4] = {CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, -1};
+#elif defined(OS_OPENBSD)
+  int mib[4] = {CTL_KERN, KERN_PROC_ARGS, getpid(), KERN_PROC_ARGV};
   size_t namesz = sizeof(name);
+  char* bin[namesz];
 
-  auto ret = sysctl(mib, 4, name, &namesz, nullptr, 0);
+  auto ret = sysctl(mib, 4, bin, &namesz, nullptr, 0);
   if (-1 == ret) {
     return nullptr;
   } else {
+    snprintf(name, sizeof(name), "%s", realpath(bin[0], nullptr));
```

Comment:
I think we should handle `realpath()` failures, which might happen in case of (1) working directory changed and bin[0] is a relative path, or (2) command was invoked without its full path and relied on PATH env var to find it. (1) is kind of tricky - unless I'm missing something, it seems we might determine the wrong path in case the relative path can also be found under the changed working directory.

Also if we continue using `realpath()`, we should handle successes differently by freeing the string it returns.

What do you think about just printing `bin[0]` directly? Do we need the exact executable path for anything?

Reply:
Actually, using bin[0] directly should just work in this case.

Reply:
I'll make the changes this weekend.